ts/Series/TemperatureMap/TemperatureMapSeries.ts

```diff
@@ -0,0 +1,445 @@
+                    }
+                });
+            });
+        }
+    }
+
+    /**
+     * Draw all the bubbles
+     * @private
+     */
+    public drawPoints(): void {
+        const series = this,
+            options = series.options,
+            temperatureColors = series.options.temperatureColors
+                .slice().reverse(); // loop form backward (easier calculations)
+
+        temperatureColors.forEach((
+            color: ColorType|[number, ColorType],
+            colorIndex: number
+        ): void => {
+            // Save size factor for further calculations
+            if (isArray(color)) { // if stops defined
+                series.sizeFactor = color[0];
+
+                color = color[1];
+            } else { // If size decreases evenly
+                series.temperatureColorIndex = colorIndex;
+            }
+
+            const colorsLength = temperatureColors.length,
+                fillColor = {
+                    radialGradient: {
+                        cx: 0.5,
+                        cy: 0.5,
+                        r: 0.5
+                    },
+                    stops: [
+                        [colorIndex === colorsLength - 1 ? 0 : 0.5, color],
+                        [1, (new Color(color)).setOpacity(0).get('rgba')]
+                    ]
+                };
+
+            options.marker = merge(options.marker, { fillColor });
+
+            series.getRadii(); // recalculate radii
+
+            series.adjustMinSize = true; // recalculate minPxSize only here
+            series.translateBubble(); // use radii
+            series.adjustMinSize = false; // reset
+
+            if (options.marker.enabled !== false || series._hasPointMarkers) {
+                const points = series.points,
+                    chart = series.chart,
+                    markerGroup = (
+                        (series as any)[series.specialGroup as any] ||
+                        series.markerGroup
+                    );
+
+                points.forEach((point): void => {
+                    point.graphics = point.graphics || [];
+
+                    let graphic = point.graphics[colorIndex];
+
+                    const pointMarkerOptions = point.marker || {},
+                        hasPointMarker = !!point.marker,
+                        verb = graphic ? 'animate' : 'attr',
+                        shouldDrawMarker = (
+                            (
+                                (options.marker || {}).enabled &&
+                                typeof pointMarkerOptions.enabled ===
+                                    'undefined'
+                            ) || pointMarkerOptions.enabled
+                        ) &&
+                        !point.isNull &&
+                        point.visible !== false &&
+                        // Above zThreshold
+                        (point.options as any).z >= (this.options.zMin || 0);
```

Comment:
The only one `as any` I couldn't get rid of. Any suggestions?

Reply:
`point.options.z` is optional, so you need to test if it is a number too. Also make sure it works with array point configs. Maybe you should check `point.z` instead?

Reply:
Oops, I saw a different error, not sure what happened. Of course it's optional. Fixed now, thanks!
ps. works fine with array point and `series.keys` config